<!DOCTYPE html>
<html>
  <head>
    <meta content="text/html;charset=utf-8" http-equiv="Content-Type"/>
    <link rel="preconnect" href="https://fonts.googleapis.com">
    <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
    <link href="https://fonts.googleapis.com/css2?family=IBM+Plex+Mono:wght@300;400;600;700&family=Lato:wght@700&display=swap" rel="stylesheet">
    <link rel="stylesheet" href="style.css"/>
  </head>
  <body>
  </script>
    <div class="grid" id="grid">
      <div class="viewport upper" id="browser-upper">
      </div>
      <div class="viewport lower" id="browser-lower">
      </div>
      <div class="sidebar upper" id="sidebar-gb1">
        <div class="payload">position: <span id="position-gb1"></span></div>
        <table>
          <thead>
            <tr>
              <th>track id</th>
              <th>offset</th>
              <th>size</th>
              <th>on/off</th>
            </tr>
          </thead>
          <tbody id="track-metadata-gb1">
          </tbody>
        </table>
      </div>
      <div class="sidebar lower" id="sidebar-gb2">
        <div class="payload">position: <span id="position-gb2"></span></div>
        <table>
          <thead>
            <tr>
              <th>track id</th>
              <th>offset</th>
              <th>size</th>
              <th>on/off</th>
            </tr>
          </thead>
          <tbody id="track-metadata-gb2">
          </tbody>
        </table>
      </div>
    </div>
    <script type="module">
      import { GenomeBrowser, default as init } from './pkg/peregrine_generic.js';
      let gb = window.gb = {};

      init().then(function() {
        gb.gb1 = new GenomeBrowser();
        gb.gb1.go({
          //backend_url: "http://staging-2020.ensembl.org/api/browser/data",
          backend_url: "http://localhost:3333/api/data http://localhost:3334/api/data",
          target_element_id: 'browser-upper'
        });
        gb.gb2 = new GenomeBrowser();
        gb.gb2.go({
          backend_url: "http://localhost:3333/api/data http://localhost:3334/api/data",
          target_element_id: 'browser-lower'
        });

        gb.gb1.tracks_on = {};
        gb.gb2.tracks_on = {};

        /* Helper functions */

        function toggle_track(genome_browser, track_name) {
          let cur = genome_browser.tracks_on[track_name];
          if(cur!==true && cur!==false) { cur = true; }
          let value = !cur;
          genome_browser.tracks_on[track_name] = value;
          let track_path = ["track"];
          // hack: uuid track id => assume expansion track
          if(track_name.length == 36) track_path.push("expand");
          track_path.push(track_name);
          genome_browser.switch(track_path, value);
        }

        function delay(t, v) {
          return new Promise(resolve => setTimeout(resolve, t, v));
        }

        function init_tracks(genome_browser, track_ids) { //enable "built-in" track(s)
          if(typeof(track_ids) == "string") track_ids = [track_ids];
          track_ids.forEach(track_id => {
            genome_browser.switch(["track", track_id], true);
            genome_browser.switch(["track", track_id, "name"], true);
            genome_browser.switch(["track", track_id, "label"], true); //used for some tracks
            if(track_id == "focus") genome_browser.switch(["track", "focus", "item"], true);
          });
        }

        function init_expansion_track(genome_browser, track_id) {
          genome_browser.switch(["track", "expand", track_id], true);
          genome_browser.switch(["track", "expand", track_id, "name"], true);
          genome_browser.switch(["track", "expand", track_id, "label"], true);
        }

        function init_variant_track(genome_browser, track_id) { //turn on all variant track switches
          init_expansion_track(genome_browser, track_id);
          genome_browser.switch(["track", "expand", track_id, "label-snv-id"], true);
          genome_browser.switch(["track", "expand", track_id, "label-snv-alleles"], true);
          genome_browser.switch(["track", "expand", track_id, "label-other-id"], true);
          genome_browser.switch(["track", "expand", track_id, "label-other-alleles"], true);
          genome_browser.switch(["track", "expand", track_id, "show-extents"], true);
        }
        
        /* Construct tracks table */
        for(const name of Object.keys(gb)){
          break;
          genome_browser = gb[name];
          let all_tracks = genome_browser.all_tracks = {};
          let position_span = document.getElementById("position-" + name);
          genome_browser.set_message_reporter(function(kind,...more) {
            if(kind=="error") {
              console.error(name + " gb error:", more[0]);
            } else if(kind=="current_position") {
              let currentPosition = more[0];
              position_span.innerText = `${currentPosition.stick}:${currentPosition.start}-${currentPosition.end}`;
            } else if(kind=="endstops") {
              console.log(name + " endstops: " + more[0])
            } else if(kind=="track_summary") {
              let tbody = document.getElementById("track-metadata-" + name);
              for(const track of more[0].summary) {
                if(track.type != "track") { continue; } /* don't care */
                all_tracks[track["switch-id"]] = track;
              }
              tbody.innerHTML = "";
              for(const track_name in all_tracks) {
                let track = all_tracks[track_name];
                if(track.type != "track") { continue; } /* don't care */
                let tr = document.createElement("tr");
                let data = [track["switch-id"],track.offset,track.height];
                for(const text of data) {
                  let td = document.createElement("td");
                  td.innerText = text;
                  tr.appendChild(td);
                }
                let td = document.createElement("td");
                td.innerHTML = "<a title='Toggle track'>üëÅÔ∏è</a>";
                td.onclick = function() { toggle_track(gb, track['switch-id']) };
                tr.append(td);
                tbody.appendChild(tr);
              }
            }
          });
        }

        
      
        /* Set initial tracks/settings */
        gb.gb1.switch(["ruler"], true);
        gb.gb1.switch(["ruler", "one_based"], true);
        init_tracks(gb.gb1, ["gene-pc-fwd","gc","focus"]);
        gb.gb1.switch(["settings","tab-selected"], "G");
        init_expansion_track(gb.gb1, "d3c13017-9f70-4d64-87a2-51cb20f042d4"); //variation track
        gb.gb1.jump("focus:gene:a7335667-93e7-11ec-a39d-005056b38ce3:ENSG00000139618"); //BRCA2

        gb.gb1.switch(["ruler"], true);
        gb.gb1.switch(["ruler", "one_based"], true);
        init_tracks(gb.gb2, ["gene-pc-fwd","gc"]);
        gb.gb2.set_stick("a7335667-93e7-11ec-a39d-005056b38ce3:13");
        gb.gb2.goto(159980000,160000500);

        //Human
        //Check track ID: curl "https://staging-2020.ensembl.org/api/tracks/track_categories/a7335667-93e7-11ec-a39d-005056b38ce3"
        // | jq -r '.track_categories[].track_list[] | select(.label == "dbSNP short variants") | .track_id'
        //Check datafiles: https://eu-west-2.console.aws.amazon.com/s3/buckets/ensembl-2020-gb-flatfiles?prefix=a7335667-93e7-11ec-a39d-005056b38ce3%2F
        //jump_variant("rs1557469781", 1, 1176408);
        //goto: position
        //genome_browser.set_stick("a7335667-93e7-11ec-a39d-005056b38ce3:13");
        //genome_browser.goto(159980000,160000500);
        //goto: variant
        //jump_variant("rs1557469781", 1, 1176408);
        //goto: gene
        //genome_browser2.jump("focus:gene:a7335667-93e7-11ec-a39d-005056b38ce3:ENSG00000139618"); //BRCA2
        //Pig
        //init_expansion_track("83090e95-2586-451c-bf3c-8ffc50068e2b"); //repeats
        //genome_browser.set_stick("da38d82b-df50-419b-af74-886463b7bfa3:14");
        //genome_browser.goto(10150673, 10329385);
      });
    </script>
  </body>
</html>
